rate_test.go

```diff
@@ -109,6 +108,98 @@ func TestRedisIsDown(t *testing.T) {
 	}
 }
 
+func TestVerify(t *testing.T) {
+	l := rateLimiter()
+
+	rate, reset, allow := l.Verify("test_verify_id", 1, time.Minute)
+	if !allow {
+		t.Fatalf("rate limited with rate %d", rate)
+	}
+	if rate != 0 {
+		t.Fatalf("got %d, wanted 1", rate)
+	}
+	dur := time.Duration(reset-time.Now().Unix()) * time.Second
+	if dur > time.Minute {
+		t.Fatalf("got %s, wanted <= %s", dur, time.Minute)
+	}
+
+	l.Allow("test_verify_id", 1, time.Minute)
+	l.Allow("test_verify_id", 1, time.Minute)
+
+	rate, reset, allow = l.Verify("test_verify_id", 1, time.Minute)
+	if allow {
+		t.Fatalf("should rate limit with rate %d", rate)
+	}
+	if rate != 2 {
+		t.Fatalf("got %d, wanted 1", rate)
+	}
+	dur = time.Duration(reset-time.Now().Unix()) * time.Second
+	if dur > time.Minute {
+		t.Fatalf("got %s, wanted <= %s", dur, time.Minute)
+	}
+}
+
+func TestVerifyMinute(t *testing.T) {
+	l := rateLimiter()
+
+	rate, reset, allow := l.VerifyMinute("test_verify_minute_id", 1)
+	if !allow {
+		t.Fatalf("rate limited with rate %d", rate)
+	}
+	if rate != 0 {
+		t.Fatalf("got %d, wanted 1", rate)
+	}
+	dur := time.Duration(reset-time.Now().Unix()) * time.Second
+	if dur > time.Minute {
+		t.Fatalf("got %s, wanted <= %s", dur, time.Minute)
+	}
+
+	l.AllowMinute("test_verify_minute_id", 1)
+	l.AllowMinute("test_verify_minute_id", 1)
+
+	rate, reset, allow = l.VerifyMinute("test_verify_minute_id", 1)
+	if allow {
+		t.Fatalf("should rate limit with rate %d", rate)
+	}
+	if rate != 2 {
+		t.Fatalf("got %d, wanted 1", rate)
+	}
+	dur = time.Duration(reset-time.Now().Unix()) * time.Second
+	if dur > time.Minute {
+		t.Fatalf("got %s, wanted <= %s", dur, time.Minute)
+	}
+}
+
+func TestVerifyHour(t *testing.T) {
+	l := rateLimiter()
+
+	rate, reset, allow := l.VerifyHour("test_verify_hour_id", 1)
+	if !allow {
+		t.Fatalf("rate limited with rate %d", rate)
+	}
+	if rate != 0 {
+		t.Fatalf("got %d, wanted 1", rate)
+	}
+	dur := time.Duration(reset-time.Now().Unix()) * time.Second
+	if dur > time.Hour {
+		t.Fatalf("got %s, wanted <= %s", dur, time.Hour)
+	}
+
+	l.AllowHour("test_verify_hour_id", 1)
+	l.AllowHour("test_verify_hour_id", 1)
+
+	rate, reset, allow = l.VerifyHour("test_verify_hour_id", 1)
+	if allow {
+		t.Fatalf("should rate limit with rate %d", rate)
+	}
+	if rate != 2 {
+		t.Fatalf("got %d, wanted 1", rate)
+	}
+	dur = time.Duration(reset-time.Now().Unix()) * time.Second
+	if dur > time.Hour {
+		t.Fatalf("got %s, wanted <= %s", dur, time.Hour)
+	}
+}
```

Comment:
These 3 tests are too verbose... can you try to replace them with table-driven tests? https://github.com/golang/go/wiki/TableDrivenTests